src/radial/radial_solver.hpp

```diff
@@ -145,517 +191,397 @@ class Radial_solver
-            sq_alpha_half = 0;
-        }
-
-        double ll_half = l__ * (l__ + 1) / 2.0;
+        struct params
+        {
+            int ir;
+            double x0;
+            double enu;
+            int l;
+            double kappa;
+            int zn;
+            Spline<double> const* ve;
+            Spline<double> const* chi_p;
+            Spline<double> const* chi_q;
+        };
 
-        double kappa{0};
-        if (rel == relativity_t::dirac) {
-            if (k__ == l__) {
-                kappa = k__;
-            } else if (k__ == l__ + 1) {
-                kappa = -k__;
-            } else {
-                std::stringstream s;
-                s << "wrong k : " << k__ << " for l = " << l__;
-                RTE_THROW(s);
+        auto func = [](double x, double const* y, double* f, void* params__) -> int {
+            using namespace radial_solver_local;
+
+            params* p = static_cast<params*>(params__);
+            double dx = x - p->x0;
+            /* reconstruct V(x) = -z/x + Vel(x) */
+            double V     = p->ve->operator()(p->ir, dx) - p->zn / x;
+            double chi_p = p->chi_p->operator()(p->ir, dx);
+            double chi_q = p->chi_q->operator()(p->ir, dx);
+            /* system of two coupled equations
+             * -------------------------------
+             *
+             *   Schroedinger radial equation case:
+             *     p'(x) = 2 * q(x) + p(x) / x
+             *     q'(x) = (V(x) - enu + l * (l + 1) / 2 / x^2) * p(x) - q(x) / x + chi_q(x)
+             *
+             *   Scalar-relativistic (Koelling-Harmon) case:
+             *     p'(x) = 2 * M(x) * q(x) + p(x) / x + chi_p(x)
+             *     q'(x) = (V(x) - enu + l * (l + 1) / 2 / M(x) / x^2) * p(x) - q(x) / x + chi_q(x)
+             *
+             *     where M(x) = 1 + 0.5 * alpha^2 * (enu - V(x))
+             *     with M = 1 (alpha -> 0) and chi_p(x) = 0 we obtain "classical" Schroedinger radial equation
+             *
+             *   ZORA case:
+             *     same as Koelling-Harmon but with M(x) = 1 - 0.5 * alpha^2 * V(x)
+             *
+             *   Dirac radial equation case:
+             *     p'(x) = alpha * (enu - V(x) + 2 * E0) * q(x) - p(x) * kappa / x
+             *     q'(x) = -alpha * (enu - V(x)) * p(x) + q(x) * kappa / x
+             */
+            switch (rel) {
+                case relativity_t::none:
+                case relativity_t::koelling_harmon:
+                case relativity_t::zora: {
+                    double ll_half = p->l * (p->l + 1) / 2.0;
+                    double M       = rel_mass<rel>(p->enu, V);
+                    /* p'(x) = 2 * M(x) * q(x) + p(x) / x + chi_p(x) */
+                    f[0] = 2 * M * y[1] + y[0] / x + chi_p;
+                    /* q'(x) = (V(x) - enu + l * (l + 1) / 2 / M(x) / x^2) * p(x) - q(x) / x + chi_q(x) */
+                    f[1] = (V - p->enu + ll_half / M / x / x) * y[0] - y[1] / x + chi_q;
```

Comment:
I may replace` /M / x / x` with `/ (M * x * x)` or better `/ (M * (x * x))`. Reason we fix the order of the operations.

Reply:
and the division is costly.